docs/pages/guides/replicating-onchain-state.mdx

```diff
@@ -0,0 +1,126 @@
+
+## Recreating onchain state
+
+We’ll use TypeScript to prototype a state representation of the bytes on chain as we receive different storage events. To make our life easier, we'll set up a little bit of boilerplate and helpers first.
+
+```ts
+type Hex = `0x${string}`;
+
+type Record = {
+  staticData: Hex;
+  encodedLengths: Hex;
+  dynamicData: Hex;
+};
+
+const store = new Map<string, Record>();
+
+// Create a key string from a table ID and key tuple to use in our store Map above
+function storeKey(tableId: Hex, keyTuple: Hex[]): string {
+  return `${tableId}:${keyTuple.join(",")}`;
+}
+
+// Like `Array.splice`, but for strings of bytes
+function bytesSplice(data: Hex, start: number, deleteCount = 0, newData: Hex = "0x"): Hex {
+  const dataNibbles = data.replace(/^0x/, "").split("");
+  const newDataNibbles = newData.replace(/^0x/, "").split("");
+  return `0x${dataNibbles
+    .splice(start, deleteCount * 2)
+    .concat(newDataNibbles)
+    .join("")}`;
+}
+
+function bytesLength(data: Hex): number {
+  return data.replace(/^0x/, "").length / 2;
+}
+```
+
+Let’s assume we’re iterating over the logs in a block. We’ll handle each event in a conditional. Setting a record is easy, because we can just replace the entire record with the arguments from the event log.
+
+```ts
+if (log.eventName === "Store_SetRecord") {
+  const key = storeKey(log.args.tableId, log.args.keyTuple);
+  store.set(key, {
+    staticData: log.args.staticData,
+    encodedLengths: log.args.encodedLengths,
+    dynamicData: log.args.dynamicData,
+  });
+}
+```
+
+Likewise, deleting a record is easy.
+
+```ts
+if (log.eventName === "Store_DeleteRecord") {
+  const key = storeKey(log.args.tableId, log.args.keyTuple);
+  store.delete(key);
+}
+```
+
+The splice events are modeled after JavaScript’s `Array.splice` and we can use our `bytesSplice` method above to simplify handling the event.
+
+```ts
+if (log.eventName === "Store_SpliceStaticData") {
+  const key = storeKey(log.args.tableId, log.args.keyTuple);
+  const record = store.get(key) ?? { staticData: "0x", encodedLengths: "0x", dynamicData: "0x" };
+  store.set(key, {
+    staticData: bytesSplice(record.staticData, log.args.start, bytesLength(log.args.data), log.args.data),
+    encodedLengths: record.encodedLengths,
+    dynamicData: record.dynamicData,
+  });
+}
+```
+
+Splicing dynamic data is roughly the same.
+
+```ts
+if (log.eventName === "Store_SpliceDynamicData") {
+  const key = storeKey(log.args.tableId, log.args.keyTuple);
+  const record = store.get(key) ?? { staticData: "0x", encodedLengths: "0x", dynamicData: "0x" };
+  store.set(key, {
+    staticData: record.staticData,
+    encodedLengths: log.args.encodedLengths,
+    dynamicData: bytesSplice(record.dynamicData, log.args.start, log.args.deleteCount, log.args.data),
+  });
+}
+```
+
+And that’s it! We’ve got our onchain state represented in TypeScript, reconstructed via MUD’s Store event logs.
```

Comment:
Maybe show the `console.log` of what the state might look like? I find it helpful to have tangible examples

Reply:
it's all just hex, so I'm not sure how useful that would be 🙈 and I am hoping most readers can understand/imagine what a splice operation does?

Reply:
I've reworded this and now link out to the other docs pages that provide more detail on the encoding.